### PR TITLE
native splitter

### PR DESCRIPTION
**Pull request by bekauz**
base implementation for #91 
to be reviewed after #98 is merged

at the moment the split configuration expects all receiver shares to add up to 100 (indicating percentages). feels like this can be simplified to take in arbitrary amounts that would get converted to fractions relative to inputs. e.g.: given three receivers with shares of `(20, 30, 10)` that add up to 60, we would be looking at a split of (20/60, 30/60, 10/60). wasn't sure if it's the right move yet so keeping it open for future.

### CODE REVIEW

File: contracts/native-splitter/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+            let splits = SPLIT_CONFIG_MAP.load(
+                deps.storage,
+                remote_chain_info.denom.to_string()
+            )?;
+
+            let mut outputs: Vec<Output> = Vec::new();
+            for split_receiver in splits.iter() {
+                // get the fraction dedicated to this receiver
+                let amt = amount
+                    .checked_multiply_ratio(split_receiver.share, Uint128::new(100));
```
Comment (Art3miX) [inline contracts/native-splitter/src/contract.rs:139]: I think using a `.map_err()?` here is nicer then a match

File: contracts/native-splitter/src/msg.rs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+#[cw_serde]
+pub struct NativeDenomSplit {
+    /// denom to be distributed
+    pub denom: String,
+    /// denom receivers and their respective shares
+    pub receivers: Vec<SplitReceiver>,
+}
+
+impl NativeDenomSplit {
+    pub fn validate(self) -> Result<NativeDenomSplit, StdError> {
```
Comment (Art3miX) [inline contracts/native-splitter/src/msg.rs:52]: Dont forget to confirm that the denoms are the same as well.

File: contracts/native-splitter/src/msg.rs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        fmt.write_str("]")?;
+        Ok(())
+    }
+}
+#[clocked]
+#[cw_serde]
+pub enum ExecuteMsg {}
+
+#[cw_serde]
+pub struct SplitConfigMap {
```
Comment (Art3miX) [inline contracts/native-splitter/src/msg.rs:98]: What is this?

**Review by Art3miX — APPROVED** [review]
Regarding the precision of the % we want, right now you only handle rounded % (50%, 51%, 25%), but there might be cases where we want a fractioned % (33.33%, 12.5%), in that case we better move to a BPS system.

I think we should stick with 100%, and not complicate it further.

File: contracts/native-splitter/src/msg.rs (hunk 3 above)
Comment (bekauz) [reply contracts/native-splitter/src/msg.rs:98]: Map that maps denoms to vec of receivers with their shares for that denom. Added a comment.

Comment (Art3miX) [reply contracts/native-splitter/src/msg.rs:98]: Are you using that? Seems like theres a state with same name?

**Pull request merged**